Atomic.md

```diff
@@ -0,0 +1,253 @@
+        in .staging
+
+-   We crash between mkdir() and InitializeEmptyPlaceholder
+
+    -   An abandoned directory is in .staging that can be safely
+        overwritten/deleted when the enumeration is retried
+
+-   One of the SetBitInFileFlags calls fails in
+     InitializeEmptyPlaceholder
+
+    -   directory is marked as in virtualization root but not empty.
+
+        -   An incomplete placeholder is in .staging that can be safely
+            overwritten/deleted when the enumeration is retried
+
+    -   directory is marked as empty but not in virtualization root
+
+        -   An incomplete placeholder is in .staging that can be safely
+            overwritten/deleted when the enumeration is retried
+
+    -   The rename() to copy the completed placeholder from .staging into
+        the virtualization root fails
+
+        -   We can retry from this error since nothing ever appeared in
+            the virtualization root. The placeholder will be overwritten
+            as part of the retry
+
+WritePlaceholderFile:
+
+-   Fopen() fails to create a file
+
+    -   We can simply retry and have this succeed. Nothing was created
+        in .staging
+
+-   Ftruncate fails to expand the file to the correct size (or is never
+     reached because of a crash)
+
+    -   We can simply retry and have this succeed. The file written in
+        .staging will be overwritten/deleted and recreated
+
+-   Memcpy'ing the providerID xattr fails (or is never reached because
+     of a crash)
+
+    -   We have an incomplete placeholder file in .staging that will be
+        overwritten by a retry.
+
+-   Memcpy'ing the contentID xattr fails (or is never reached because of
+     a crash)
+
+    -   We have an incomplete placeholder file in .staging that will be
+        overwritten by a retry.
+
+-   We crash before reaching InitializeEmptyPlaceholder
+
+    -   We have a partially created placeholder written into .staging that
+        will be overwritten on retry
+
+-   One of the SetBitInFileFlags calls fails in
+     InitializeEmptyPlaceholder
+
+    -   Directory is marked as in virtualization root but not empty.
+
+        -   An incomplete placeholder is in .staging that can be safely
+            overwritten/deleted when the enumeration is retried
+
+-   Chmod() fails or is never reached due to a crash
+
+    -   We can retry and have this succeed as the placeholder hasn't
+        left .staging yet
+
+    -   The rename() to copy the completed placeholder from .staging into
+        the virtualization root fails
+
+        -   We can retry from this error since nothing ever appeared in
+            the virtualization root. The placeholder will be overwritten
+            as part of the retry
+
+With this approach, each placeholder file that will be created will be moved in one at a time to their final location rather than moving all placeholders that may be present in a directory at once. 
```

Comment:
Will we create/move each placeholder, or first create all and then move all? Is there a meaningful difference between the two?

Reply:
We will create and move each placeholder incrementally rather than move all placeholders at once. I added some more content to why we didn't choose rename_xp() based designs.

Reply:
Yes I understand that we can't move them "all at once", meaning atomically. I was more wondering, should the flow be: first create all the placeholders, one at a time; then move all the placeholders, one at a time.

The one advantage I can think of is that we at least got as far as _creating_ all of the placeholders, before we attempt to move any of them into the actual virtualization root, so there's less chance of needing to clean anything up later.